Implement max connections

DarrenTsung opened the pull request:
Add `max_connections` to `hyper::Client` builder + keep track of connections through `hyper::Client::Pool`.

If there too many connections, sending a new HTTP/1 request will not create a new connection, but will instead wait for an idle connection (if possible). Otherwise the request will fail.

Commit:
Add idle_connection_max and don't make new connection if past max

Commit:
Rename to max_idle_connections

Commit:
Use generic parameter for more ergonomic API

Commit:
Use raii_counter to count all connections instead of just idle

Counts connections that are Connecting, Active (Pooled), and Idle.
Http/2 connections are shared between requests, so we keep a Counter
on the Idle and don't count the Pooled types handed out.

coder543 commented on src/client/pool.rs:
```diff
@@ -60,6 +62,12 @@ struct Connections<T> {
+    // This is used to keep a maximum bound on the number of connections
+    // This is a WeakCounter so it is not counted towards the Conns count.
+    conns_count: WeakCounter,
+    // A maximum number of idle connections
```
is it a maximum number of idle connections, or total connections? I thought the goal was to prevent a massive spike in number of total connections under heavy load, where the connections aren't idle.

DarrenTsung replied:
Total connections - I think this comment is just old

DarrenTsung replied:
Done

coder543 commented on src/client/mod.rs:
```diff
@@ -313,7 +316,8 @@ where C: Connect + Sync + 'static,
                 // 1. Connect is canceled if this is HTTP/2 and there is
-                //    an outstanding HTTP/2 connecting task.
+                //    an outstanding HTTP/2 connecting task OR this is HTTP/1
+                //    and there are too many idle connections.
```
same thing here

DarrenTsung replied:
Done

Commit:
Fix max_connections comment

Commit:
Update max_connections doc comment for Builder

Commit:
Fix old comment - too many connections

Commit:
Update Http1TooManyIdle error to TooManyConnections

Commit:
Evict idle connections if limited by max_connections

Commit:
Revert "Evict idle connections if limited by max_connections"

This reverts commit 420dca639c2e848e8001da51cc4abedda7507619.

Commit:
Allow external user to sync Counter with conns_counter

instead of throttling from `hyper::Client`, information is just exposed
now.

Commit:
Error logs when conn is closed

Commit:
Logging around connection closing

Commit:
Log removing connections which keys

Commit:
Add error logging for pooling idle conn

Commit:
Race is won when finished connecting, instead of connect + handshake

This should reduce connection thrash by reducing the amount of
connections dropped by winning the race.

Commit:
Remove waiter logic, should not interrupt connect unless first poll

Commit:
Remove pooling idle connection log

Commit:
Don't run checkout (even if fail) if connect finishes first

Commit:
Revert "Remove waiter logic, should not interrupt connect unless first poll"

This reverts commit 9164cbf5889629457b9d741c2931761b30151c4f.

Commit:
Checkout returns Idle connection or else None
